server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +270,181 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
     return paContinue;
 }
 
-void SC_PortAudioDriver::GetPaDeviceFromName(const char* device, int* mInOut) {
-    const PaDeviceInfo* pdi;
-    const PaHostApiInfo* apiInfo;
-    char devString[256];
-    PaDeviceIndex numDevices = Pa_GetDeviceCount();
-    mInOut[0] = paNoDevice;
-    mInOut[1] = paNoDevice;
+std::string SC_PortAudioDriver::GetPaDeviceName(int index) const {
+    auto* pdi = Pa_GetDeviceInfo(index);
+    std::string name;
+#ifndef __APPLE__
+    name += Pa_GetHostApiInfo(pdi->hostApi)->name;
+    name += " : ";
+#endif
+    name += pdi->name;
+    return name;
+}
+
+void SC_PortAudioDriver::GetPaDeviceFromName(const char* device, int* mInOut, IOType ioType) const {
+    if (device[0] == '\0')
+        return;
 
-    // This tries to find one or two devices that match the given name (substring)
-    // might cause problems for some names...
+    PaDeviceIndex numDevices = Pa_GetDeviceCount();
+    *mInOut = paNoDevice;
     for (int i = 0; i < numDevices; i++) {
-        pdi = Pa_GetDeviceInfo(i);
-        apiInfo = Pa_GetHostApiInfo(pdi->hostApi);
-        strcpy(devString, apiInfo->name);
-        strcat(devString, " : ");
-        strcat(devString, pdi->name);
-        if (strstr(devString, device)) {
-            if (pdi->maxInputChannels > 0)
-                mInOut[0] = i;
-            if (pdi->maxOutputChannels > 0)
-                mInOut[1] = i;
+        auto* pdi = Pa_GetDeviceInfo(i);
+        std::string devString = GetPaDeviceName(i);
+        if (strstr(devString.c_str(), device)) {
+            if (ioType == IOType::Input && pdi->maxInputChannels > 0) {
+                *mInOut = i;
+                break;
+            } else if (ioType == IOType::Output && pdi->maxOutputChannels > 0) {
+                *mInOut = i;
+                break;
+            }
         }
     }
 }
 
+PaStreamParameters SC_PortAudioDriver::GetPaStreamParameters(int device, int channelCount,
+                                                             double suggestedLatency) const {
```

Comment:
You always pass 0 for `suggestedLatency` -- you can remove this parameter and hard-code it, I think. Also, I generally prefer "MakeXYZ" for functions that are essentially wrappers for object construction.

Reply:
For this one I'd like to leave `suggestedLatency` in - we'll use it once this function is used in the refactor in #4477

Reply:
I changed the name though